include/Mat.h

```diff
@@ -254,35 +290,62 @@ class Mat {
         buildStrides();
      
         data = new Type[size()];
+        base = newMat(data);
         size_type i = 0;
         for(auto elem : list){
             data[i] = elem;
             i++;
         }
+    }
+
+    template<typename... arg>
+    Mat(std::initializer_list<Type> list, AllocInfo<Type> alloc, const arg... ind){
+        ndim = sizeof...(arg);
+        if(ndim > 32) throw invalid_argument("too many dimensions (<=32 allowed)");
+
+        size_type temp_dims[32]{(static_cast<size_type>(ind))...};
+        size_type temp_strides[32];
+
+        temp_strides[ndim-1] = 1;
+        for(long j = ndim-2; j >= 0; j--){
+            temp_strides[j] = temp_strides[j+1]*temp_dims[j+1];
+        }
+
+        allocator = alloc;
+        DimInfo dInfo(ndim, temp_dims, temp_strides);
+        allocator->allocate(this, alloc.userdata, dInfo);
+
+        size_type i = 0;
+        for(auto elem : list){
+            data[i] = elem;
+            i++;
+        }
+
         base = newMat(data);
     }
 
     Mat(const Mat& b){
-        if(b.base != NULL) base = b.base;
+        if(b.base != NULL) base = b.base; //remove this later when new base type is added
         else base = const_cast<Mat<Type>*>(&b);
         (base->refCount)++;
 
         ndim = b.ndim;
         data = b.data;
         
         if(b.allocator != NULL){
-            DimInfo dInfo(ndim, dims, strides);
-            b.allocator->allocate(this, dInfo);
+            allocator = b.allocator;
+            DimInfo dInfo(ndim, b.dims, b.strides);
+            allocator->allocate(this, allocator->userdata, dInfo);
```

Comment:
Make sure to document that allocate is responsible for detecting the difference between an initial allocation and the copy constructor. Put this doc where DimInfo is defined